### PR TITLE
map editor: check hero and town spells during mod assessment

### PR DESCRIPTION
**Pull request by godric3**
While looking at some test campaigns (like one from #4376) I noticed that town possibleSpells list have all spells from all mods active while adding town. Map creator wasn't aware that there is some reference to mod in his map so it wasn't marked as required. This PR should both add error on validation and improve marking required mods based on map objects

### CODE REVIEW

File: mapeditor/mapcontroller.cpp [hunk 1]
```diff
@@ -613,16 +613,49 @@ ModCompatibilityInfo MapController::modAssessmentAll()
...
 {
 	ModCompatibilityInfo result;
+
+	auto extractEntityMod = [&result](const auto & entity) 
+	{
+		auto modName = QString::fromStdString(entity->getJsonKey()).split(":").at(0).toStdString();
+		if(modName != "core")
+			result[modName] = VLC->modh->getModInfo(modName).getVerificationInfo();
+	};
+
```
Comment (IvanSavenko) [inline mapeditor/mapcontroller.cpp:623]: Perhaps instead we should add `getModScope` method to entity class?
More clear and `getJsonKey()` is often implemented as `return modScope + ':' + identifier;`

File: lib/CPlayerState.cpp [hunk 2]
```diff
@@ -50,6 +50,11 @@ std::string PlayerState::getJsonKey() const
 	return color.toString();
 }
 
+std::string PlayerState::getModScope() const
+{
+	return "core";
```
Comment (godric3) [inline lib/CPlayerState.cpp:55]: I guess core make sense here as it's not something provided by mods

File: lib/ObstacleHandler.cpp [hunk 3]
```diff
@@ -31,6 +31,11 @@ std::string ObstacleInfo::getJsonKey() const
 	return identifier;
 }
 
+std::string ObstacleInfo::getModScope() const
+{
+	return "core";
```
Comment (godric3) [inline lib/ObstacleHandler.cpp:36]: `ObstacleInfo` is missing `modScope` but I think obstacles can be provided by mods. Is it intentional or should I introduce it and initialize in `loadFromJson`?

File: mapeditor/mapcontroller.cpp [hunk 4]
```diff
@@ -613,16 +613,49 @@ ModCompatibilityInfo MapController::modAssessmentAll()
...
 	for(auto obj : map.objects)
 	{
-		if(obj->ID == Obj::HERO)
-			continue; //stub! 
-		
 		auto handler = obj->getObjectHandler();
-		auto modName = QString::fromStdString(handler->getJsonKey()).split(":").at(0).toStdString();
-		if(modName != "core")
-			result[modName] = VLC->modh->getModInfo(modName).getVerificationInfo();
+		auto modScope = QString::fromStdString(handler->getJsonKey()).split(":").at(0).toStdString();
```
Comment (godric3) [inline mapeditor/mapcontroller.cpp:627]: does it make sense to introduce `getModScope` to `AObjectTypeHandler`?

File: lib/ObstacleHandler.cpp (hunk 3 above)
Comment (IvanSavenko) [reply lib/ObstacleHandler.cpp]: Yeah, sounds like a bug.
While not 100% safe - fixing this one might reveal bugs someplace else (if any), but still a good idea to do.

**Pull request merged**